tinygrad/runtime/ops_metal.py

```diff
@@ -1,15 +1,58 @@
 from __future__ import annotations
-import os, subprocess, pathlib, ctypes, tempfile, functools
-import Metal, libdispatch
-from typing import List, Any, Tuple, Optional
-from tinygrad.helpers import prod, getenv, DEBUG, unwrap2
-from tinygrad.device import Compiled, Compiler, CompileError, LRUAllocator
+import os, subprocess, pathlib, tempfile, functools
+from typing import List, Any, Tuple, Optional, cast
+from tinygrad.helpers import prod, getenv, DEBUG
+from tinygrad.device import Compiled, Compiler, LRUAllocator
 from tinygrad.renderer.cstyle import MetalRenderer
+import ctypes
+
+class objc_id(ctypes.c_void_p): # This prevents ctypes from converting response to plain int, and dict.fromkeys() can use it to dedup
+  def __hash__(self): return hash(self.value)
+  def __eq__(self, other): return self.value == other.value
+
+class objc_instance(objc_id): # method with name "new", "alloc" should be freed after use
+  def __del__(self): msg(self, "release")
+
+def load_library(path: str): return ctypes.CDLL(path)
+
+libobjc = load_library("/usr/lib/libobjc.dylib")
+libmetal = load_library("/Library/Frameworks/Metal.framework/Metal")
+# Must be loaded for default Metal Device: https://developer.apple.com/documentation/metal/1433401-mtlcreatesystemdefaultdevice?language=objc
+load_library("/Library/Frameworks/CoreGraphics.framework/CoreGraphics")
+libdispatch = load_library("/usr/lib/libSystem.dylib") # libdispatch is part of libSystem on mac
+libobjc.objc_getClass.restype = objc_id
+libobjc.sel_registerName.restype = objc_id
+libmetal.MTLCreateSystemDefaultDevice.restype = objc_instance
+libdispatch.dispatch_data_create.restype = objc_instance
+
+def msg(ptr: objc_instance, selector: str, /, *args: Any, restype: type = objc_id) -> Any:
+  sender = libobjc["objc_msgSend"] # Using attribute access returns a new reference so setting restype is safe
+  sender.restype = restype
+  return sender(ptr, libobjc.sel_registerName(selector.encode()), *args)
+
+def to_ns_str(s: str) -> objc_instance: return msg(libobjc.objc_getClass(b"NSString"), "stringWithUTF8String:", s.encode())
+
+def to_ns_array(items: List[Any]): return (objc_instance * len(items))(*items)
```

Comment:
I don't think this function does what it's name says, and maybe you don't need it at all